compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -368,6 +370,7 @@ private void analyzeTopLevelNodes(BLangPackage pkgNode, SymbolEnv pkgEnv) {
     }
 
     private void analyzeNode(BLangNode node, SymbolEnv env) {
+        logErrorsForInferredArrays(node);
```

Comment:
Can't we avoid this here completely and do the validation when visiting `BLangArrayType`?

Reply:
If we do so, we can't validate the return type, since `BLangArrayType` visitor isn't called when there is an array as the return type. Is there any workaround for that?

Reply:
Or we can add a method like `validateReturnType` in `BLangFunction` and `BLangFunctionTypeNode` visitors to validate the return type and remove this `logErrorsForInferredArrays`?

Reply:
Let's try and restructure this as discussed offline, to add the missing type node visits and do the validation in the visit for `BLangArrayType`.